fm.1
.TH FM 1 "fm 1.0.2"
.SH NAME
fm \- control frequency, volume, mute/unmute of FM radio card
.SH SYNOPSIS
.HP
.B fm
[
.B \-h
] [
.B \-o
] [
.B \-q
] [
.B \-d
.I device
] [
.B \-t
.I tuner
] [
.B \-T
.B none
|
.B forever
|
.I time
]
.B on
|
.B off
|
.B +
|
.B \-
|
.B freq
[
.I volume
]
.SH DESCRIPTION
.B fm
is a program to control the frequency, volume, and mute/unmute state
of an FM radio card, using the video4linux interface introduced in
2.1.x series Linux kernels.
.SS OPTIONS
.PP
.TP
.B \-h
Print a usage message to standard output, and exit.
.TP
.B \-o
Ignore frequency range limits of card.  Use if you suspect that the
card supports a wider frequency range than its driver believes.
.TP
.B \-q
Quiet mode.  Keeps information on station and volume from being
printed on standard output.
.TP
\fB\-d \fIdevice
Sets \fIdevice\fR as the device to tune.  The default is
\fB/dev/radio0\fR.
.TP
\fB\-t \fItuner
Sets \fItuner\fR as the tuner on the selected device to adjust.  The
default is tuner 0.  Most radio devices have only a single tuner.
.TP
\fB\-T none \fR| \fBforever \fR| \fItime
After tuning, sleep for the time specified or forever.  Time is
specified in seconds by default.  A suffix of
.B m
indicates minutes,
.B h
indicates hours, or
.B d
indicates days.
.IP
The
.B -T
option is useful with radio card drivers that only maintain the tuner
settings while the tuner's file descriptor remains open.  Try using
this option if running
.B fm
ordinarily produces only a single "pop" from your speakers.
.TP
.BI freq
Frequency to tune the radio to, in MHz.  For instance, \fB88.9\fR
specifies a frequency of 88.9 MHz.  AM tuner values are also specified
in MHz; for instance, 530 kHz would be specified as \fB.530\fR.
.TP
.BI on
Turn the radio on (unmute).
.TP
.BI off
Turn the radio off (mute).
.TP
.BI volume
Specify the desired volume, in percent.  Not all radio devices support
volume control.
.TP
.BI +
Increase the current volume.
.TP
.BI -
Decrease the current volume.
.SH CONFIGURATION
.B fm
reads \fI$HOME\fB/.fmrc\fR, if it exists, to obtain default settings.
Each line may take one of the following forms:
.TP
.BR VOL \ percent
Specifies default volume (default: 12.5%).
.TP
.BR INCR \ percent
Volume increment used for
.B +
and
.B -
options (default: 10%).
.TP
.BR TIME
Default sleep time (default:
.BR none ).
.PP
All other lines are ignored.
.SH SEE ALSO
.TP
Additional documentation:
.B /usr/share/doc/fmtools/README
.TP
The fmtools homepage:
.B http://benpfaff.org/fmtools
.SH AUTHORS
Russell Kroll <rkroll@exploits.org>, now maintained by Ben Pfaff
<blp@cs.stanford.edu>.  Sleep time feature contributed by Dave Ulrick
<d-ulrick@insightbb.com>.  This manpage written by Ben Pfaff.


NEWS
2.0.1:

Sat Jan 9 13:28:53 PST 2010 / Ben Pfaff <blp@cs.stanford.edu>

- Fix bugs that prevented fm and fmscan from working outside "test"
  mode (used by the testsuite.

- Better support radios that do not have volume control.

- Make fm report when the radio is muted and tell the user how to unmute it.

- Packaging improvements.

2.0:

Sat Dec 12 21:18:52 PST 2009 / Ben Pfaff <blp@cs.stanford.edu>

- Rewrite to use video4linux2 API.

- Add "-q" (quiet) and "-T" (tuner) options to fmscan.

- Improve fmscan frequency calculations.  It should now be better at
  accurately finding radio stations.

- Add Autotest test suite and, in the process, convert the packaging
  to use Autoconf and Automake.

- Some code and some documentation used a range of 0-65535 for volume,
  other code and other documentation used a range of 0-100.  This has
  been consistently changed to 0-100.

1.0.2:

Thu Nov 23 1:41:00 PDT 2006 / Ben Pfaff <blp@cs.stanford.edu>

- Fix version number reported by fm and fmscan programs, as well as
  the version number included in the manpages.

  Bug report by "Petr Slansky" <slansky@usa.net>.

- No functional changes.

1.0.1:

Sat Oct 7 1:05:00 PDT 2006 / Ben Pfaff <blp@cs.stanford.edu>

- Make fm able to sleep for a user-specified time, or forever, after
  applying settings.  This makes fm usable with radio tuner drivers
  that turn off the radio after the fd used for tuning is closed.

  Idea and patch from Dave Ulrick <d-ulrick@insightbb.com>.

1.0:

Sat Sep 3 12:21:00 PDT 2004 / Ben Pfaff <blp@cs.stanford.edu>

- New maintainer.  New home page is http://benpfaff.org/fmtools

- Updated email addresses.

- Integrated Debian packaging.

- Made fmscan threshold configurable from command line.

- Use videodev.h header detached from kernel to avoid including kernel
  headers.

0.99.1:

Fri Aug 16 23:08:30 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - fm, fmscan: fix the FSF's address in the headers

Fri Aug 16 23:07:49 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - COPYING: put in the modern version of the GPL v2, with the correct
   address for the FSF.

Fri Aug 16 23:05:33 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - Makefile: reworked a bunch of stuff, and added such useful things
   like install targets.  Hopefully this will make packaging less of 
   a hack.

Fri Aug 16 22:55:41 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - fm: do the VIDIOCGAUDIO first so there's always good data in the struct

Fri Aug 16 22:48:10 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - fm.1: incorporated the man page written by Ben Pfaff.

Fri Aug 16 22:47:01 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - fm: set va.balance to 32768 rather than relying on whatever junk
   happens to be in the struct.   Reported by ahmet inan.

Fri Aug 16 22:43:47 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - fm: applied patch from Viktor Rosenfeld to fix the rounding so that
   channels like 99.95 MHz can be tuned.

0.99.0:

Sat Feb  3 22:41:12 CST 2001 - Russell Kroll <rkroll@exploits.org>

 - Final cleanups for 0.99.0.  This source has been sitting around for
   awhile, so it's probably time to push towards 1.0 and call it done.
   How much can you do to a pair of radio programs, anyway?

Thu Jan  4 17:29:49 CST 2001 - Russell Kroll <rkroll@exploits.org>

 - fmscan: ioctl/code cleanups, ala fm
 - fmscan: -d <dev> support
 - fm: -d <dev> support for multiple devices (/dev/radio1 and so on)
 - fm: ioctl results checked, other coding cleanups

0.2.5 and earlier:

Mon Apr 12 20:32:13 MDT 1999
 - -t option for selecting the tuner on multi-tuner boards (radio-cadet, etc)

Wed Feb  3 13:13:01 MST 1999
 - fmtools 0.2.4 released
 - 'devices' target added to Makefile
 - documentation updated for fmscan and /dev/radio0
 - programs now use /dev/radio0 instead of old(er) /dev/radio
 - fmscan added
 - small code cleanups
 - -o added to oneliner usage display

Sat Jan 16 23:55:15 MST 1999
 - fmtools 0.2.3 released
 - -o (override frequency range checks) introduced after some l-k discussion

Sun Jan 10 22:17:07 MST 1999
 - fmtools 0.2.2 released
 - overflows and negative values during volume changes fixed
 - exit codes are now sane
 - the division should now accurately reflect 100% volume

Thu Jul 30 08:39:50 MDT 1998
 - We now support fine tuning on cards that fly the VIDEO_TUNER_LOW flag.
   Thanks to Ben Pfaff for this addition.


fm.c
/* fm.c - simple V4L2 compatible tuner for radio cards

   Copyright (C) 2004, 2006, 2009 Ben Pfaff <blp@cs.stanford.edu>
   Copyright (C) 1998 Russell Kroll <rkroll@exploits.org>

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2 of the License, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
   more details.

   You should have received a copy of the GNU General Public License along with
   this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <math.h>
#include <errno.h>
#include <stdio.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <limits.h>

#include "fmlib.h"

#define DEFAULT_DEVICE	"/dev/radio0"

static double
clamp(double percent)
{
        return percent < 0.0 ? 0.0 : percent > 100.0 ? 100.0 : percent;
}

static int
convert_time (const char *string)
{
	if (strcmp(string, "forever") == 0 ||
            strcmp(string, "-") == 0 ||
            atoi(string) < 0)
		return 0;
	else if (strcmp(string, "none") == 0 ||
                 strcmp(string, "0") == 0)
		return -1;
	else
	{
		char worktime[80+1];
		int inttime;
		const char *suffix;

		suffix = string + strspn(string, "0123456789");

		strncpy(worktime, string, suffix - string);
		worktime[suffix - string] = '\0';
		inttime = atoi(worktime);

		switch (*suffix)
		{
		case 's':
		case '\0':
			break;
		case 'm':
			inttime *= 60;
			break;
		case 'h':
			inttime *= 60 * 60;
			break;
		case 'd':
			inttime *= 24 * 60 * 60;
			break;
		default:
			break;
		}

		return inttime;
	}
}



static char *
format_time (char *buffer, const char *string)
{
        if (strcmp(string, "forever") == 0 ||
            strcmp(string, "-") == 0 ||
            atoi(string) < 0)
                strcpy(buffer, "forever");
        else if (strcmp(string, "none") == 0 ||
                 strcmp(string, "0") == 0)
                strcpy(buffer, "none");
        else
        {
                char worktime[80+1];
                const char *suffix;
                char *format;
                int int_time;

                suffix = string + strspn(string, "0123456789");
                strncpy(worktime, string, suffix - string);
                worktime[suffix - string] = '\0';
                int_time = atoi(worktime);

                switch (*suffix)
                {
                case 'm':
                        format = "%d minute(s)";
                        break;
                case 'h':
                        format = "%d hour(s)";
                        break;
                case 'd':
                        format = "%d day(s)";
                        break;
                case 's':
                case '\0':
                default:
                        format = "%d second(s)";
                break;
                }

                sprintf(buffer, format, int_time);
        }

        return buffer;
}

static void
maybe_sleep(const struct tuner *tuner, const char *wait_time)
{
        char message[80+1];
        int int_wait_time;

        int_wait_time = convert_time(wait_time);

        if (int_wait_time > 0)
        {
                printf("Sleeping for %s\n", format_time(message, wait_time));
                tuner_sleep(tuner, int_wait_time);
        } else if (int_wait_time == 0)
        {
                printf("Sleeping forever...CTRL-C exits\n");
                for (;;)
                        pause();
        }
}

static void
usage(void)
{
	printf("fmtools fm version %s\n\n", VERSION);
	printf("usage: %s [-h] [-o] [-q] [-d <dev>] [-t <tuner>] "
               "[-T none|forever|time] <freq>|on|off [<volume>]\n\n", 
               program_name);
	printf("A small controller for Video for Linux radio devices.\n\n");
	printf("  -h         display this help\n");
	printf("  -o         override frequency range limits of card\n");
	printf("  -q         quiet mode\n");
	printf("  -d <dev>   select device (default: /dev/radio0)\n");
	printf("  -t <tuner> select tuner (default: 0)\n");
        printf("  -T <time>  after setting frequency, sleep for some time\n"\
               "             (default: none; -=forever)\n");
	printf("  <freq>     frequency in MHz (i.e. 94.3)\n");
	printf("  on         turn radio on\n");
	printf("  off        turn radio off (mute)\n");
	printf("  +          increase volume\n");
	printf("  -          decrease volume\n");
	printf("  <volume>   percentage (0-100)\n");
	exit(EXIT_SUCCESS);
}

static void
getconfig(const char *fn,
          double *defaultvol, double *increment, char *wait_time)
{
	FILE	*conf;
	char	buf[256];

        if (!fn) {
                snprintf(buf, sizeof buf, "%s/.fmrc", getenv("HOME"));
                fn = buf;
        }
	conf = fopen(fn, "r");

	if (!conf)
		return;

	while(fgets(buf, sizeof(buf), conf)) {
		buf[strlen(buf)-1] = 0;
		if (!strncmp(buf, "VOL", 3))
			sscanf(buf, "%*s %lf", defaultvol);
		if (!strncmp(buf, "INCR", 3))
			sscanf(buf, "%*s %lf", increment);
		if (!strncmp(buf, "TIME", 4))
			sscanf(buf, "%*s %s", wait_time);
	}

	fclose(conf);
}

static void
print_volume(const struct tuner *tuner)
{
        if (tuner_has_volume_control(tuner))
                printf(" at %.2f%% volume\n", tuner_get_volume(tuner));
        else
                printf(" (radio does not support volume control)\n");
}

int main(int argc, char **argv)
{
        struct tuner tuner;
	const char *device = NULL;
        int index = 0;
        bool quiet = false;
        bool override = false;
        const char *config_file = NULL;
	double defaultvol = 12.5;
	double increment = 10;
        char wait_time_buf[256+1] = "";
	const char *wait_time = NULL;

        program_name = argv[0];

	/* need at least a frequency */
	if (argc < 2)
		fatal(0, "usage: %s [-h] [-o] [-q] [-d <dev>] [-t <tuner>] "
                      "[-T time|forever|none] "
                      "<freq>|on|off [<volume>]\n", program_name);

        for (;;) {
                int option = getopt(argc, argv, "+qhot:T:c:d:");
                if (option == -1)
                        break;

		switch (option) {
                case 'q':
                        quiet = 1;
                        break;
                case 'o':
                        override = 1;
                        break;
                case 't':
                        index = atoi(optarg);
                        break;
                case 'T':
                        wait_time = optarg;
                        break;
                case 'd':
                        device = optarg;
                        break;
                case 'c':
                        config_file = optarg;
                        break;
                case 'h':
                default:
                        usage();
                        break;
		}
	}

	getconfig(config_file, &defaultvol, &increment, wait_time_buf);
        if (!wait_time)
                wait_time = *wait_time_buf ? wait_time_buf : "none";

	argc -= optind;
	argv += optind;

	if (argc == 0)		/* no frequency, on|off, or +|- given */
		usage();

        tuner_open(&tuner, device, index);

        if (!strcmp(argv[0], "off")) {
                tuner_set_mute(&tuner, true);
                if (!quiet)
                        printf("Radio muted\n");
        } else if (!strcmp(argv[0], "on")) {
                tuner_set_mute(&tuner, false);
                if (!quiet) {
                        printf("Radio on");
                        print_volume(&tuner);
                }
        } else if (!strcmp(argv[0], "+") || !strcmp(argv[0], "-")) {
                double new_volume;
                if (!tuner_has_volume_control(&tuner))
                        fatal(0, "Radio does not support volume control");

                new_volume = tuner_get_volume(&tuner);
                if (argv[0][0] == '+')
                        new_volume += increment;
                else
                        new_volume -= increment;
                new_volume = clamp(new_volume);

                if (!quiet)
                        printf("Setting volume to %.2f%%\n", new_volume);

                tuner_set_volume(&tuner, new_volume);
        } else if (atof(argv[0])) {
                double frequency = atof(argv[0]);
                double volume = argc > 1 ? clamp(atof(argv[1])) : defaultvol;
                tuner_set_freq(&tuner, frequency * 16000.0, override);
                tuner_set_volume(&tuner, volume);
                if (!quiet) {
                        printf("Radio tuned to %2.2f MHz", frequency);
                        print_volume(&tuner);
                }
        } else {
                fatal(0, "unrecognized command syntax; use --help for help");
        }
        maybe_sleep(&tuner, wait_time);
        tuner_close(&tuner);
	return 0;
}
